### PR TITLE
feat(cli): cli interface

### LINKED ISSUE #35: Create a CLI to wrap the framework-core
The framework-core library should have a command-line interface so users can more easily integrate it with other environments just by downloading the node package and executing it i.e. GitHub Actions.

Given the bash command `code-suggester`, create arguments for:
- upstream-repo
- upstream-owner
- pr-description
- pr-title
- branch-name
- force
- files
- git-dir

as discussed in [go/auto-pr-framework](go/auto-pr-framework)

### PR DESCRIPTION
**Pull request by TomKristie**
Create a CLI interface to interact with the framework-core
It:

- takes arguments and maps them to user PR options
- takes a git directory and generates a change set Map from the git diff items
- initializes octokit and a logger
- invokes framework-core
- includes core-logic testing

Closes #35 

Redo of #51

### CODE REVIEW

File: src/bin/code-suggester.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  .check(argv => {
+    for (const key in argv) {
+      if (typeof argv[key] === 'string' && !argv[key]) {
+        throw Error(
+          `String parameters cannot be provided empty values. Parameter ${key} was given an empty value`
+        );
+      }
+    }
+    if (argv.files !== undefined && argv['git-dir'] !== undefined) {
+      throw Error('"git-dir" options must be defined');
```
Comment (chingor13) [inline src/bin/code-suggester.ts:99]: From the previous PR review: files option was removed.

File: src/bin/handle-git-dir-change.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+// See https://git-scm.com/docs/git-diff#Documentation/git-diff.txt-git-diff-filesltpatterngt82308203
+type GitDiffStatus = 'A' | 'D' | 'M' | 'T' | 'U' | 'X' | string;
+
+/**
+ * Get the absolute path of a relative path
+ * @param {string} dir the wildcard directory containing git change, not necessarily the root git directory
+ * @returns {string} the absolute path relative to the path that the user executed the bash command in
+ */
+export function resolvePath(dir: string) {
```
Comment (chingor13) [inline src/bin/handle-git-dir-change.ts:35]: This function doesn't seem to be adding much, any reason we shouldn't just use path.resolve wherever this is being used?

File: src/bin/handle-git-dir-change.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+/**
+ * Get the git root directory.
+ * @param {string} dir the wildcard directory containing git change, not necessarily the root git directory
+ * @returns {string} the absolute path of the git directory root
+ */
+export function findRepoRoot(dir: string): string {
+  try {
+    return execSync('git rev-parse --show-toplevel', {cwd: dir})
+      .toString()
+      .slice(0, -1); // remove the \n
```
Comment (chingor13) [inline src/bin/handle-git-dir-change.ts:49]: Is there a string.chomp() or something a bit safer to remove the trailing newline character(s)?

File: src/bin/handle-git-dir-change.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+            });
+          }
+        );
+      }
+    } catch (err) {
+      logger.warning(
+        `\`git diff --raw\` may have changed formats: \n ${gitDiffPattern}`
+      );
+      reject(err);
+    }
```
Comment (chingor13) [inline src/bin/handle-git-dir-change.ts:103]: A catch block around a long section of code makes it difficult to see which error is being handled. If this is to catch the parsing error, try to wrap just the parsing with the try block.

File: src/bin/handle-git-dir-change.ts [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+ * @returns {string[]} a list of git diffs
+ */
+export function getAllDiffs(gitRootDir: string): string[] {
+  execSync('git add -A', {cwd: gitRootDir});
+  const diffs: string[] = execSync('git diff --raw --staged --no-renames', {
+    cwd: gitRootDir,
+  })
+    .toString() // strictly return buffer for mocking purposes. sinon ts doesn't infer {encoding: 'utf-8'}
+    .split('\n'); // remove the trailing new line
+  diffs.pop();
```
Comment (chingor13) [inline src/bin/handle-git-dir-change.ts:119]: split + pop seems like an odd way to trim the trailing newline.

Comment (TomKristie) [reply src/bin/handle-git-dir-change.ts:119]: Oh that comment is for line 119. We want to split because we are given a list of stuff separated by `\n`
i.e. `'hello\n\world\foo\n`.split('\n') = ['hello', 'world', 'foo', '']`
the pop removes the extra line

**Comment by TomKristie** [comment]
I added an explicit git installation check which I didn't test since I consider that an integration test which is outside of scope. Let me know what you think! @chingor13

File: src/bin/handle-git-dir-change.ts (hunk 2 above)
Comment (TomKristie) [reply src/bin/handle-git-dir-change.ts:42]: I separated it out so that I can explicitly test that absolute paths are generated the way I want to.

File: src/bin/code-suggester.ts (hunk 1 above)
Comment (TomKristie) [reply src/bin/code-suggester.ts]: I deleted the whole conditional out since yargs `demandOption: true` is enabled

File: src/bin/handle-git-dir-change.ts [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+ * Rejects if there is a git diff error, or if the file contents could not be loaded.
+ * @param {string} gitRootDir the root of the local GitHub repository
+ * @param {string} gitDiffPattern A single file diff. Renames and copies are broken up into separate diffs. See https://git-scm.com/docs/git-diff#Documentation/git-diff.txt-git-diff-filesltpatterngt82308203 for more details
+ * @returns {Promise<GitFileData>} the current mode, the relative path of the file in the Git Repository, and the file status.
+ */
+export function getGitFileData(
+  gitRootDir: string,
+  gitDiffPattern: string
+): Promise<GitFileData> {
+  return new Promise((resolve, reject) => {
```
Comment (chingor13) [inline src/bin/handle-git-dir-change.ts:105]: Would async/await be easier here?

File: test/fixtures/fixtures [hunk 7]
```diff
@@ -0,0 +1 @@
+build/test/fixtures/
```
Comment (chingor13) [inline test/fixtures/fixtures:1]: Is this file necessary?

File: test/git-dir-handler.ts [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+import * as child_process from 'child_process';
+
+before(() => {
+  setup();
+});
+
+// tslint:disable:no-unused-expression
+// .null triggers ts-lint failure, but is valid chai
+// .true triggers ts-lint failure, but is valid chai
+describe('git directory diff output to git file data + content', () => {
```
Comment (chingor13) [inline test/git-dir-handler.ts:36]: nit: usually the describe block shows the function name under test
```suggestion
describe('getGitFileData', () => {
```

File: src/bin/handle-git-dir-change.ts [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+  }
+}
+
+/**
+ * Get the absolute path of a relative path
+ * @param {string} dir the wildcard directory containing git change, not necessarily the root git directory
+ * @returns {string} the absolute path relative to the path that the user executed the bash command in
+ */
+export function resolvePath(dir: string) {
+  const absoluteDir = path.resolve(process.cwd(), dir);
```
Comment (chingor13) [inline src/bin/handle-git-dir-change.ts:43]: If this is only used for the CLI then implicitly depending on the cwd may be ok. If this will be included later in the core (which I could potentially see), we instead want the relative root to be explicit (and provided by the caller - e.g. the CLI).

File: src/bin/handle-git-dir-change.ts (hunk 6 above)
Comment (TomKristie) [reply src/bin/handle-git-dir-change.ts:105]: I felt it was easier to explicitly be able to get the `readfile` error for clearer error messaging. Otherwise I'd have to have nested try catch blocks if i wanted more precise error logging.

File: src/bin/handle-git-dir-change.ts (hunk 9 above)
Comment (TomKristie) [reply src/bin/handle-git-dir-change.ts:43]: this is explicitly for the CLI. I've intended the bin directory to be strictly for the CLI. If we want to do work for the framework-core it will not be handled here. It is a relatively small function any way.

File: test/fixtures/fixtures (hunk 7 above)
Comment (TomKristie) [reply test/fixtures/fixtures]: No. This must have been acidental!!

File: src/bin/handle-git-dir-change.ts [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+    );
+  }
+}
+
+/**
+ * Load the change set asynchronously.
+ * @param dir the directory containing git changes
+ * @returns {Promise<Changes>} the change set
+ */
+export function getChanges(dir: string): Promise<Changes> {
```
[the remark this answers is not in the document]
Comment (TomKristie) [reply src/bin/handle-git-dir-change.ts:208]: Not necessarily. It's only required when we have an `await` statement.

File: src/bin/handle-git-dir-change.ts [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+  try {
+    validateGitInstalled();
+    const absoluteDir = resolvePath(dir);
+    const gitRootDir = findRepoRoot(absoluteDir);
```
[the remark this answers is not in the document]
Comment (TomKristie) [reply src/bin/handle-git-dir-change.ts:212]: I'll create an issue
